### PR TITLE
Fixes #1051 Remove need for Addressable

### LINKED ISSUE #1051: Look into refactoring away from Addressable
In the interests of reducing our dependency list, I'd like to research if we can utilize Ruby's URI class directly rather than relying on [Addressable](https://github.com/sporkmonger/addressable)

https://docs.ruby-lang.org/en/3.2/URI.html

### PR DESCRIPTION
**Pull request by anantone**
- replace with equivalent method from built-in URI class
- create two _ad how_ methods in `utils.rb` where broader character class needed

This is a 🙋 feature or enhancement. 

## Summary

Below is a replacement table for the former calls to Addressable. The two new methods, along with needed constants for character classes, are in `utils.rb`. 

| Addressable method  | Used in  |  URI equivalent  |
| --- | --- | --- |
| Addressable::URI.encode | bridgetown-routes/lib/roda/plugins/bridgetown_routes.rb | new method broad_uri_encode(str) |
| Addressable::URI.escape | bridgetown-core/lib/bridgetown-core/collection.rb | (alias for Addressable::URI.encode) |
|  Addressable::URI.unencode | bridgetown-core/lib/bridgetown-core/utils.rb |URI.decode_uri_component(str) |
| Addressable::URI.unescape | bridgetown-core/lib/bridgetown-core/model/plugin_origin.rb (2x), bridgetown-core/lib/bridgetown-core/model/repo_origin.rb | (alias for Addressable::URI.unencode) |
| Addressable::URI.parse | bridgetown-core/lib/bridgetown-core/utils.rb, bridgetown-core/lib/bridgetown-core/filters/url_filters.rb (4x), bridgetown-core/lib/bridgetown-core/resource/destination.rb | URI.parse |
| Addressable::URI.normalize_component | bridgetown-core/lib/bridgetown-core/filters.rb, bridgetown-core/lib/bridgetown-core/generated_page.rb | new normalize_component method  | 

## Context

Fixes #1051 (or at least tries to contribute! this would be my first contribution) 
I have added a new test, but in all honesty, I don't know if I've covered all that the changes entail. I'll add comments to the PR where I still have doubts. I hope this can facilitate resolving the referenced issue, and would be grateful for any help in getting there, if at all possible.

### CODE REVIEW

File: bridgetown-core/lib/bridgetown-core/utils.rb [hunk 1]
```diff
@@ -10,6 +10,27 @@ module Utils # rubocop:todo Metrics/ModuleLength
...
+    ALPHA = "a-zA-Z"
+    DIGIT = "0-9"
+    GEN_DELIMS = "\\:\\/\\?\\#\\[\\]\\@"
+    SUB_DELIMS = "\\!\\$\\&\\'\\(\\)\\*\\+\\,\\;\\="
+    RESERVED = (GEN_DELIMS + SUB_DELIMS).freeze
+    UNRESERVED = "#{ALPHA}#{DIGIT}\\-\\.\\_\\~".freeze
+    RESERVED_AND_UNRESERVED = RESERVED + UNRESERVED
+
+    # URI encode with broader set of character classes (same as Addressable)
+    def broad_uri_encode(str)
```
Comment (anantone) [inline bridgetown-core/lib/bridgetown-core/utils.rb]: This method was initially called `addressable_encode`, since that is what it replaces, but I changed it to a more descriptive name (reflecting "broader character classes for uri encoding", as in broader than the built-in URI class provides for). Happy to consider better names, though!

File: bridgetown-core/test/test_utils.rb [hunk 2]
```diff
@@ -418,4 +418,15 @@ class TestUtils < BridgetownUnitTest
       end
     end
   end
+
+  describe "The new `normalize_component` method" do
+    it "both decodes and encodes like Addressable::URI:normalize_component did" do
+      str1 = "simple%2Fex%61mple "
+      expected1 = "simple/example%20"
+      assert_equal normalize_component(str1), expected1
+      str2 = "a%25more%26complicated%24example"
```
Comment (anantone) [inline bridgetown-core/test/test_utils.rb]: I tried, but am not sure what would be most relevant to test for here!

**Comment by anantone** [comment]
I also have not removed Addressable from the Gemfile and such, until it is confirmed that it won't be needed anymore!

**Comment by jaredcwhite** [comment]
Thanks @anantone! Taking a look now.

**Comment by jaredcwhite** [comment]
@anantone Hey, so the strategy here seems sound, but I would recommend verifying your the test suite runs locally (`bundle exec rake test_all` in the project root). I saw there are a lot of missing references to `Bridgetown::Utils` (aka calling `broad_uri_encode` on its own in the `Bridgetown::Collection` class isn't going to work), and in one case there's a call to `addressable_encode` which doesn't exist at all in the repo. Did you write that code and it didn't get checked in perhaps? Anyway, would love to review again once you've fixed those errors. TY!

**Comment by anantone** [comment]
@jaredcwhite Thank you for your patience! This new commit fixes the most obvious mistakes that you pointed out (including the non existing `addressable_encode` method, which should have been replaced by then!). But there are still a lot of test failures related to URI formats, so I'll need some time (and probably help) to figure that out...

**Comment by jaredcwhite** [comment]
@anantone No problem! Happy to help troubleshoot some of the test cases if you run into a blocker.

**Comment by anantone** [comment]
Thanks! Well, I did some testing today, and there are A LOT of errors on the same model as this one:
```
Error:
TestFrontMatterDefaults::A site with front matter defaults with no path or type#test_0001_affects all types and paths:
URI::InvalidURIError: bad URI (is not URI?): "repo://posts.collection/_posts/2014-03-22-escape-+%20%2520[].markdown"
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:130:in `split'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:135:in `parse'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/common.rb:247:in `parse'
    lib/bridgetown-core/model/base.rb:17:in `origin_for_id'
    lib/bridgetown-core/model/base.rb:12:in `find'
    lib/bridgetown-core/collection.rb:297:in `read_resource'
    lib/bridgetown-core/collection.rb:86:in `block in read'
    lib/bridgetown-core/collection.rb:79:in `each'
    lib/bridgetown-core/collection.rb:79:in `read'
    lib/bridgetown-core/reader.rb:41:in `block in read_collections'
    lib/bridgetown-core/reader.rb:38:in `each_value'
    lib/bridgetown-core/reader.rb:38:in `read_collections'
    lib/bridgetown-core/reader.rb:22:in `read'
    lib/bridgetown-core/concerns/site/processable.rb:75:in `read'
    lib/bridgetown-core/concerns/site/processable.rb:16:in `process'
    test/test_front_matter_defaults.rb:172:in `block (2 levels) in <class:TestFrontMatterDefaults>'
```
I asked some AI friends who suggested that there was an interaction with the slugify method which is right there in utils.rb. Now I tried some of the suggested fixes from AI friends, but those didn't work (as is mostly the case, in my experience). But it does seem that there is an unintended error being introduced with URI formats, although I don't know why because I don't yet see the overall picture, in which slugify seems to play a large role.

**Comment by fpsvogel** [comment]
@jaredcwhite Outside this PR @anantone and I have been corresponding about this. I suggested this issue to him as a good first issue, and he did a great job investigating it, but it has turned out to be unexpectedly complex.

I pushed up some commits to address some of the missing pieces, but some tests still fail.

Even if all tests were passing, though, I wouldn't feel confident that we've replicated the behavior in `Addressable::URI`. The behaviors in Ruby's `URI` differ from `Addressable::URI` in ways that we'd need more edge case tests to discover.

Also, `Addressable::URI` is more forgiving with input to decoding and parsing, whereas `URI` throws an error for special characters that are not encoded. I added some preemptive encoding to prevent those errors ([1](https://github.com/bridgetownrb/bridgetown/pull/1063/files#diff-ec4cad0e03427eaa7c4deeaf33bb4cba9471e824690ee841f31113c957a27bb0R41), [2](https://github.com/bridgetownrb/bridgetown/pull/1063/files#diff-ec4cad0e03427eaa7c4deeaf33bb4cba9471e824690ee841f31113c957a27bb0R63)), but that seems like a bad solution because if the input is already partially encoded, then encoding it again will mess up the input (because encoding is not idempotent, e.g. `URI.encode_uri_component("%20")` returns `"%2520"`).

Plus, to encode Unicode characters as Punycode (as in [this test case](https://github.com/bridgetownrb/bridgetown/blob/main/bridgetown-core/test/test_filters.rb#L686-L693)), we would need to pull in a gem, maybe https://github.com/skryukov/uri-idna

All that to say, I'm starting to wonder whether we're taking the right approach in trying to replicate `Addressable::URI` behaviors one-to-one. Of course, if we didn't aim for that, breaking changes would be likely. But with the present approach, bugs seem very likely, and it might make sense to stick with `Addressable::URI` if we want the behavior we've been getting from `Addressable::URI` 🤷

**Comment by jaredcwhite** [comment]
@fpsvogel Let's abort then, imo this issue isn't that important relative to a lot of other things in our backlog right now. I was just hoping to reduce our dependency count if it's an easy win.

@anantone Sorry to have had you spinning your wheels on this, but I hope it was a valuable learning experience! Please feel free to poke away on other things as you may find worthwhile.

**Comment by anantone** [comment]
@jaredcwhite Not a problem, it was indeed a great learning experience for me (thanks again @fpsvogel) And after all, the issue asked to "look into" refactoring away from Addressable, which we did! 

One last contribution after poking around at 3 am (just me, no AI friends this time): one example of what seemed like a simple substitution but isn't, and one example where a similar "trick" works fine.
## substituting `URI.parse` for `Addressable::URI.parse`
got the following five errors:
```
Error:
TestFilters::filters::absolute_url filter#test_0014_does not raise a TypeError when passed a hash:
URI::InvalidURIError: bad URI (is not URI?): "{\"foo\"=>\"bar\"}"
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:130:in `split'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:135:in `parse'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/common.rb:247:in `parse'
    lib/bridgetown-core/filters/url_filters.rb:91:in `compute_absolute_url'
    lib/bridgetown-core/filters/url_filters.rb:14:in `absolute_url'
    test/test_filters.rb:707:in `block (3 levels) in <class:TestFilters>'

Error:
TestFilters::filters::absolute_url filter#test_0011_normalizes international URLs:
URI::InvalidURIError: URI must be ascii only "http://\u00FCmlaut.example.org//"
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:84:in `split'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:135:in `parse'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/common.rb:247:in `parse'
    lib/bridgetown-core/filters/url_filters.rb:97:in `compute_absolute_url'
    lib/bridgetown-core/filters/url_filters.rb:14:in `absolute_url'
    test/test_filters.rb:692:in `block (3 levels) in <class:TestFilters>'

Error:
TestFilters::filters::relative_url filter#test_0004_normalizes international URLs:
URI::InvalidURIError: URI must be ascii only "\u9519\u8BEF.html"
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:84:in `split'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:135:in `parse'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/common.rb:247:in `parse'
    lib/bridgetown-core/filters/url_filters.rb:107:in `compute_relative_url'
    lib/bridgetown-core/filters/url_filters.rb:24:in `relative_url'
    test/test_filters.rb:749:in `block (3 levels) in <class:TestFilters>'

Error:
TestFilters::filters::relative_url filter#test_0013_does not normalize absolute international URLs:
URI::InvalidURIError: URI must be ascii only "https://example.com/\u9519\u8BEF"
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:84:in `split'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:135:in `parse'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/common.rb:247:in `parse'
    lib/bridgetown-core/filters/url_filters.rb:107:in `compute_relative_url'
    lib/bridgetown-core/filters/url_filters.rb:24:in `relative_url'
    test/test_filters.rb:818:in `block (3 levels) in <class:TestFilters>'

Error:
TestFilters::filters::jsonify filter#test_0004_converts drop with drops to json:
URI::InvalidURIError: bad URI (is not URI?): "http://example.com/base/+/%# +/"
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:130:in `split'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/rfc3986_parser.rb:135:in `parse'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/uri-1.1.1/lib/uri/common.rb:247:in `parse'
    lib/bridgetown-core/resource/destination.rb:20:in `absolute_url'
    lib/bridgetown-core/resource/base.rb:223:in `absolute_url'
    lib/bridgetown-core/resource/base.rb:294:in `to_h'
    lib/bridgetown-core/resource/base.rb:307:in `as_json'
    lib/bridgetown-core/resource/base.rb:311:in `to_json'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/json-2.16.0/lib/json/common.rb:441:in `generate'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/json-2.16.0/lib/json/common.rb:441:in `generate'
    lib/bridgetown-core/drops/drop.rb:145:in `to_json'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/json-2.16.0/lib/json/common.rb:443:in `generate'
    /home/anantone/.rbenv/versions/3.2.9/lib/ruby/gems/3.2.0/gems/json-2.16.0/lib/json/common.rb:443:in `generate'
    lib/bridgetown-core/drops/drop.rb:145:in `to_json'
    lib/bridgetown-core/filters.rb:158:in `jsonify'
    test/test_filters.rb:936:in `block (3 levels) in <class:TestFilters>'

925 tests, 2256 assertions, 0 failures, 5 errors, 64 skips
```

## substituting `URI.decode_uri_component` for both `Addressable::URI.unencode` and `Addressable::URI.unescape` 
passes existing tests cleanly

**Comment by fpsvogel** [comment]
@jaredcwhite Agreed, I'll go ahead and close the Issue.

@anantone Thanks, that lines up with a big challenge that I found in this refactor: several similarly-named operations between `URI` and `Addressable::URI` actually had quite different behaviors (`encode`, `parse`, `normalize`), and it isn't obvious how to make the behaviors match up when using `URI`. If we really wanted to do that thoroughly, maybe we'd basically end up implementing parts of `Addressable::URI`!

**Comment by anantone** [comment]
@fpsvogel Yes, I was starting to have the same thought: if we're trying to reproduce so much behavior that is specific to `Addressable::URI`... maybe it makes sense to just use it!

(for the record, I just tried to reproduce `Addressable::URI.parse` (by encoding non-ASCII characters before parsing, among other things), and went from 5 errors to 123 errors... So I'm good, for the time being, thank you!) :)
